### PR TITLE
generate default eslint config from vue cli

### LINKED ISSUE #597: Migrate from tslint to eslint
##### **Overview of the feature request**

Migrate linting from tslint to eslint for typescript.

##### **Motivation for or Use Case**

tslint is deprecated and eslint is recommended in its place.

##### **Related issues or PR**

-   [X] Checking this box is mandatory (this is just to show you read everything)

### LINKED ISSUE #608: Prettier / eslint: vue files
I think it would be cool to have prettier / eslint for all vue files
Probably related to https://github.com/jhipster/jhipster-vuejs/issues/597 and https://github.com/jhipster/jhipster-vuejs/pull/601

### PR DESCRIPTION
**Pull request by atomfrede**
With this PR the ap generates a default eslint config for vue + typescript as the vue-cli would create. As of our policy 1, I think thats fine. The huge work ahead to adhere to or define exceptions from the rules. 

TODOs:

* [ ] Cleanup: Remove `tslint` config

We have with a default app 

```
40 errors and 158 warnings found.
```

Which seems okay to me. Most warnings (as it seems) are from tests with not used variables.

:question: Can we delete/cleanup old files in a blueprint? So can we overwrite the `cleanup.js` in a blueprint?

Closes #597
Updates #600  

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/jhipster-vuejs/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/jhipster-vuejs/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
> Can we delete/cleanup old files in a blueprint? So can we overwrite the cleanup.js in a blueprint?

Do you think it's possible @mshima ?

File: generators/client/templates/vue/.eslintrc.js.ejs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      '@vue/prettier',
+      '@vue/prettier/@typescript-eslint'
+    ],
+    parserOptions: {
+      ecmaVersion: 2020
+    },
+    rules: {
+      'no-console': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
+      'no-debugger': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
+      '@typescript-eslint/no-explicit-any': 'off',
```
Comment (atomfrede) [inline generators/client/templates/vue/.eslintrc.js.ejs:19]: These are just temporary until we have cleaned up the templates

File: generators/client/templates/vue/src/main/webapp/app/shared/data/data-utils.service.ts.ejs [hunk 2]
```diff
@@ -78,10 +78,10 @@ export default class JhiDataUtils extends Vue {
    */
   clearInputImage(entity, elementRef, field, fieldContentType, idInput) {
     if (entity && field && fieldContentType) {
-      if (entity.hasOwnProperty(field)) {
+      if (Object.prototype.hasOwnProperty.call(entity, field)) {
         entity[field] = null;
       }
-      if (entity.hasOwnProperty(fieldContentType)) {
+      if (Object.prototype.hasOwnProperty.call(entity, fieldContentType)) {
```
Comment (atomfrede) [inline generators/client/templates/vue/src/main/webapp/app/shared/data/data-utils.service.ts.ejs:84]: These are errors from es lint, which are now fixed/

File: generators/client/templates/vue/src/main/webapp/app/shared/jhi-item-count.vue.ejs [hunk 3]
```diff
@@ -3,9 +3,11 @@
         <span v-if="i18nEnabled"
               v-text="$t('global.item-count', {first, second, total })">Showing {{first}} - {{second}} of {{total}} items.</span>
     <span v-if="!i18nEnabled">
+      <!-- eslint-disable vue/no-parsing-error -->
```
Comment (atomfrede) [inline generators/client/templates/vue/src/main/webapp/app/shared/jhi-item-count.vue.ejs:6]: This rule is disabled as I am pretty sure it is a false positive here.

**Comment by atomfrede** [comment]
@pascalgrimaud I have ignored all warnings for now (a lot of unused vars, missing type declarations etc). If you like you can give this a try. The config is aligned with the default one from vue-cli, so Policy 1 is totally in place :)

**Comment by pascalgrimaud** [comment]
Good job, I'll have a look @atomfrede 
For your information, the builds failed because of:

```
Error: Copying template vue/package.json.ejs failed. [ReferenceError: /home/runner/work/jhipster-vuejs/jhipster-vuejs/generators/client/templates/vue/package.json.ejs:163
    161|     "lint": "vue-cli-service lint",
    162|     "lint:fix": "<%= clientPackageManager %> run lint <%= optionsForwarder %>--fix",
 >> 163|     "cleanup": "rimraf <%= DIST_DIR %> <%= AOT_DIR %>",
    164|     "clean-www": "rimraf <%= DIST_DIR %>app/{src,<%= BUILD_DIR %>}",
    165|     "start": "<%= clientPackageManager %> run webpack:dev",
    166|     "start-tls": "<%= clientPackageManager %> run webpack:dev <%= optionsForwarder %>--env.tls",
```

Can you fix it ?

**Comment by atomfrede** [comment]
> Good job, I'll have a look @atomfrede
> For your information, the builds failed because of:
> 
> ```
> Error: Copying template vue/package.json.ejs failed. [ReferenceError: /home/runner/work/jhipster-vuejs/jhipster-vuejs/generators/client/templates/vue/package.json.ejs:163
>     161|     "lint": "vue-cli-service lint",
>     162|     "lint:fix": "<%= clientPackageManager %> run lint <%= optionsForwarder %>--fix",
>  >> 163|     "cleanup": "rimraf <%= DIST_DIR %> <%= AOT_DIR %>",
>     164|     "clean-www": "rimraf <%= DIST_DIR %>app/{src,<%= BUILD_DIR %>}",
>     165|     "start": "<%= clientPackageManager %> run webpack:dev",
>     166|     "start-tls": "<%= clientPackageManager %> run webpack:dev <%= optionsForwarder %>--env.tls",
> 
> AOT_DIR is not defined]
>     at module.exports.error (/home/runner/generator-jhipster/generators/generator-base.js:1590:15)
>     at /home/runner/generator-jhipster/generators/utils.js:256:23
>     at tryHandleCache (/home/runner/generator-jhipster/node_modules/ejs/lib/ejs.js:275:14)
>     at Object.exports.renderFile (/home/runner/generator-jhipster/node_modules/ejs/lib/ejs.js:478:10)
>     at Object.renderContent (/home/runner/generator-jhipster/generators/utils.js:252:9)
> ```
> 
> Can you fix it ?

I have seen it, but thats strange as I did not change something there. But I will have a look.

**Comment by kaidohallik** [comment]
I made #628 for fix. This error is caused by jhipster/generator-jhipster#11907

**Comment by atomfrede** [comment]
Thanks @kaidohallik

**Comment by mshima** [comment]
> Do you think it's possible @mshima ?

Saw just now.

> Can we delete/cleanup old files in a blueprint? So can we overwrite the cleanup.js in a blueprint?

Is there specific behavior to override cleanup.js?

I suggest to add a task to writing priority: 
```
cleanup() {
  this.deleteDestination('tslint.json');
}
```

The behavior is a bit different than the generator-base-private's removeFile:
- deleteDestination (yeoman-generator) - queues the file to be removed at the conflicts priority.
- removeFile (generator-base-private) - removes the file from filesystem with shelljs.

`isJhipsterVersionLessThan` can be used in conjunction of `this.jhipsterOldVersion = this.config.get('jhipsterVersion');` but depends on what stage the following was called.

I would suggest each blueprint to write it's own generatedWith version:
```
// Assume 6.8.0 if value is not set.
this.oldBlueprintVersion = this.blueprintConfig.get('generatedWith') || '6.8.0';
this.blueprintConfig.set('generatedWith', packageJson.version);
```

```
if (semver.lt(this.oldBlueprintVersion, desiredVersion)) {
  ... deleteFile;
}
```

**Comment by atomfrede** [comment]
Thanks @mshima for the detailed explanation. I think we can go with deleting the not needed tslint config.

@pascalgrimaud rebased against master which removes aot_dir.

**Comment by pascalgrimaud** [comment]
Just tried it, and here 2 comments:
- npm run lint -> there is an auto fix here, instead of displaying lint error, the behavior has changed here
- all vue files uses 2 indents instead of 4, so our generated files will conflict a lot when we'll upgrade our application

**Comment by atomfrede** [comment]
Both should be configurable. Will have a look.

**Comment by atomfrede** [comment]
@pascalgrimaud both should be fixed now.

**Comment by pascalgrimaud** [comment]
@atomfrede : is this PR still in draft ?

**Comment by atomfrede** [comment]
The old tslint config file is not yet removed. Otherwise it's good to be merged. The now ignored rules can be fixed afterwards imho

**Comment by atomfrede** [comment]
Regarding the cleanup we could 

* Do it here (e.g. look at the current config/package.json to determine the used blueprint version). And create a cleanup procedure like we have for the main generator
* Keep the `tslint.json` config file for now (it is not generated anymore for new apps)
* Just try to delete `tslint.json`

**Comment by pascalgrimaud** [comment]
I'm reviewing this. Here some comments:

After the initial generation, when running `npm run lint`, I have:

```
1789 warnings found.
1789 warnings potentially fixable with the `--fix` option.
```

So, I launch: `npm run lint:fix`, and I still have:

```
5 errors and 1416 warnings found.
1413 warnings potentially fixable with the `--fix` option.
```

I think it's not nice to have so much warning.
So I completed with prettier:

```diff
diff --git a/.prettierrc b/.prettierrc
index f9b9911..8e8480b 100644
--- a/.prettierrc
+++ b/.prettierrc
@@ -16,3 +16,6 @@ overrides:
   - files: "*.java"
     options:
       tabWidth: 4
+  - files: "*.vue"
+    options:
+      tabWidth: 4
diff --git a/package.json b/package.json
index cdc807e..d3fdfab 100644
--- a/package.json
+++ b/package.json
@@ -102,7 +102,7 @@
     ]
   },
   "scripts": {
-    "prettier:format": "prettier --write \"{,src/**/,webpack/}*.{md,json,js,ts,tsx,css,scss}\"",
+    "prettier:format": "prettier --write \"{,src/**/,webpack/}*.{md,json,js,ts,tsx,css,scss,vue}\"",
     "lint": "vue-cli-service lint --no-fix",
     "lint:fix": "vue-cli-service lint",
     "cleanup": "rimraf target/classes/static/",
```

Then, I needed to fix some prettier error, and finally, the result is better:

```
16 warnings found.
16 warnings potentially fixable with the `--fix` option.
```

With only 16 warnings, I'm sure we can fix all of them to have 0 warning.

@atomfrede :
- we should add prettier to help us here
- we should fix the prettier error
- we should fix the last warning
- what do you think ?

**Comment by atomfrede** [comment]
Sounds good. In my sample application I think I was already on 0 lint warnings/errors but I adapted it manually. Should we have an upgrade notice to manually have the old tslint config file deleted and come up with a general approach how to clean up old files?

**Comment by pascalgrimaud** [comment]
We can simply add it the our release note.
No need to add more code in our blueprint to delete this file...

**Comment by pascalgrimaud** [comment]
@atomfrede : do you want me to finish this PR, with my previous comments ?

**Comment by atomfrede** [comment]
@pascalgrimaud If you have some spare time please go ahead. Otherwise I will pick it up on wednesday most likely.

**Comment by pascalgrimaud** [comment]
So, I didn't manage to go further, as there is no similar method in this blueprint: https://github.com/jhipster/generator-jhipster/blob/master/generators/generator-base-private.js#L1601-L1610

File: generators/client/templates/vue/src/main/webapp/app/shared/data/data-utils.service.ts.ejs (hunk 2 above)
Comment (china-fengguan) [reply generators/client/templates/vue/src/main/webapp/app/shared/data/data-utils.service.ts.ejs:84]: could we use the following script, and remove the 'if' condition?
`Object.keys(entity).forEach`

**Comment by pascalgrimaud** [comment]
Sorry for this PR @atomfrede but is it possible to report your commits into generator-jhipster ?
I tried to finish your PR before working on the merge, but I was stuck.

**Comment by atomfrede** [comment]
Sure was already thinking if we finish it here or just in the main generator.

**Comment by pascalgrimaud** [comment]
I prefer generator-jhipster directly, as I think I won't wait months before doing v7 release.
Remember we can do v7-beta release :-)

**Comment by pascalgrimaud** [comment]
closed by https://github.com/jhipster/generator-jhipster/pull/12972

**Pull request closed without merging**